node/src/components/consensus/highway_core/evidence.rs

```diff
@@ -47,32 +77,78 @@ impl<C: Context> Evidence<C> {
     /// the two units are the same.
     pub(crate) fn validate(
         &self,
-        v_id: &C::ValidatorId,
+        validators: &Validators<C::ValidatorId>,
         instance_id: &C::InstanceId,
     ) -> Result<(), EvidenceError> {
         match self {
             Evidence::Equivocation(unit1, unit2) => {
-                if unit1.wire_unit.creator != unit2.wire_unit.creator {
-                    return Err(EvidenceError::EquivocationDifferentCreators);
+                Self::validate_equivocation(unit1, unit2, instance_id, validators)
+            }
+            Evidence::Endorsements {
+                endorsement1,
+                unit1,
+                endorsement2,
+                unit2,
+                swimlane2,
+            } => {
+                let v_id = validators
+                    .id(endorsement1.validator_idx())
+                    .ok_or(EvidenceError::UnknownPerpetrator)?;
+                if *endorsement1.unit() != unit1.hash() || *endorsement2.unit() != unit2.hash() {
+                    return Err(EvidenceError::EndorsementWrongHash);
                 }
-                if unit1.wire_unit.seq_number != unit2.wire_unit.seq_number {
-                    return Err(EvidenceError::EquivocationDifferentSeqNumbers);
+                if endorsement1.validator_idx() != endorsement2.validator_idx() {
+                    return Err(EvidenceError::EndorsementDifferentCreators);
                 }
-                if unit1.wire_unit.instance_id != *instance_id
-                    || unit2.wire_unit.instance_id != *instance_id
-                {
-                    return Err(EvidenceError::EquivocationInstanceId);
+                for (unit, pred) in iter::once(unit2).chain(swimlane2).tuple_windows() {
+                    if unit.wire_unit.previous() != Some(&pred.hash()) {
+                        return Err(EvidenceError::EndorsementInvalidSwimlane);
+                    }
                 }
-                if unit1 == unit2 {
-                    return Err(EvidenceError::EquivocationSameUnit);
-                }
-                if !C::verify_signature(&unit1.hash(), v_id, &unit1.signature)
-                    || !C::verify_signature(&unit2.hash(), v_id, &unit2.signature)
+                Self::validate_equivocation(
+                    unit1,
+                    swimlane2.last().unwrap_or(unit2),
```

Comment:
How is this `unwrap_or` ? Why not use `unit2` always?

Reply:
The last entry of the swimlane is the one that has the same sequence number as `unit1` and can thus directly be used to verify that the two forks are conflicting.

If `unit2` has the same sequence number as `unit1` already, the `swimlane2` will be empty.